packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
-    // Schedule a check so that we test this new annotation to see if any mouse
-    // is currently inside its region. It has to happen after the frame is
-    // complete so that the annotation layer has been added before the check.
-    if (mouseIsConnected) {
-      _scheduleMousePositionCheck();
+  // The collection of annotations that are currently being tracked.
+  // It is operated by [attachAnnotation] and [detachAnnotation].
+  final Set<MouseTrackerAnnotation> _trackedAnnotations = <MouseTrackerAnnotation>{};
+  bool get _hasAttachedAnnotations => _trackedAnnotations.isNotEmpty;
+
+  void _addMouseDevice(int device, PointerEvent event) {
+    final bool wasConnected = mouseIsConnected;
+    assert(!_mouseStates.containsKey(device));
+    _mouseStates[device] = _MouseState(mostRecentEvent: event);
+    // Schedule a check to enter annotations that might contain this pointer.
+    _checkDeviceUpdates(device: device);
+    if (mouseIsConnected != wasConnected) {
+      notifyListeners();
     }
   }
 
-  /// Stops tracking an annotation, indicating that it has been removed from the
-  /// layer tree.
-  ///
-  /// An assertion error will be thrown if the associated layer is not removed
-  /// and receives another mouse hit.
-  void detachAnnotation(MouseTrackerAnnotation annotation) {
-    final _TrackedAnnotation trackedAnnotation = _findAnnotation(annotation);
-    for (int deviceId in trackedAnnotation.activeDevices) {
-      if (annotation.onExit != null) {
-        final PointerEvent event = _lastMouseEvent[deviceId] ?? _pendingRemovals[deviceId];
-        assert(event != null);
-        annotation.onExit(PointerExitEvent.fromMouseEvent(event));
+  void _removeMouseDevice(int device, PointerEvent event) {
+    final bool wasConnected = mouseIsConnected;
+    assert(_mouseStates.containsKey(device));
+    final _MouseState disconnectedMouseState = _mouseStates.remove(device);
+    disconnectedMouseState.mostRecentEvent = event;
+    // Schedule a check to exit annotations that used to contain this pointer.
+    _checkDeviceUpdates(
+      device: device,
+      disconnectedMouseState: disconnectedMouseState,
+    );
+    if (mouseIsConnected != wasConnected) {
+      notifyListeners();
+    }
+  }
+
+  // Handler for events coming from the PointerRouter.
+  void _handleEvent(PointerEvent event) {
+    if (event.kind != PointerDeviceKind.mouse) {
+      return;
+    }
+    final int device = event.device;
+    if (event is PointerAddedEvent) {
+      _addMouseDevice(device, event);
+    } else if (event is PointerRemovedEvent) {
+      _removeMouseDevice(device, event);
+    } else if (event is PointerHoverEvent) {
+      final _MouseState mouseState = _guaranteeMouseState(device, event);
+      final PointerEvent previousEvent = mouseState.mostRecentEvent;
+      mouseState.mostRecentEvent = event;
+      if (previousEvent is PointerAddedEvent || previousEvent.position != event.position) {
```

Comment:
Why do we need to send a hover event if previousEvent is PointerAddedEvent, but the position hasn't changed?